zapcore/sampler.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import (
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+const (
+	_numLevels        = _maxLevel - _minLevel
+	_countersPerLevel = 4096
+)
+
+func newCounters() *counters {
+	return &counters{}
+}
+
+type counters [_numLevels][_countersPerLevel]atomic.Uint64
+
+func (c *counters) Inc(lvl Level, key string) uint64 {
+	return c.get(lvl, key).Inc()
+}
+
+func (c *counters) Reset(lvl Level, key string) {
+	c.get(lvl, key).Store(0)
+}
+
+func (c *counters) get(lvl Level, key string) *atomic.Uint64 {
+	i := lvl - _minLevel
+	j := fnv32a(key) % _countersPerLevel
```

Comment:
@prashantv confirmed that the 1.7 compiler is sufficiently clever here.

The modulo compiles to `& 4096-1`:
```
         .          .      8a701: ANDL $0xfff, AX
```

As seen in:
```
(pprof) disasm counters..get
Total: 34.89s
ROUTINE ======================== go.uber.org/zap/zapcore.(*counters).get
     1.33s      3.81s (flat, cum) 10.92% of Total
     120ms      120ms      8a6a0: GS MOVQ GS:0x8a0, CX
      10ms       10ms      8a6a9: CMPQ 0x10(CX), SP
         .          .      8a6ad: JBE 0x8a728
      80ms       80ms      8a6af: SUBQ $0x28, SP
      40ms       40ms      8a6b3: MOVQ BP, 0x20(SP)
      10ms       10ms      8a6b8: LEAQ 0x20(SP), BP
         .          .      8a6bd: MOVQ 0x40(SP), AX
      60ms       60ms      8a6c2: MOVQ AX, 0(SP)
      50ms       50ms      8a6c6: MOVQ 0x48(SP), AX
      60ms       60ms      8a6cb: MOVQ AX, 0x8(SP)
      70ms       70ms      8a6d0: MOVZX 0x38(SP), AX
      30ms       30ms      8a6d5: INCL AX
      40ms       40ms      8a6d7: MOVB AL, 0x1f(SP)
      60ms      2.54s      8a6db: CALL go.uber.org/zap/zapcore.fnv32a(SB)
      50ms       50ms      8a6e0: MOVL 0x10(SP), AX
     240ms      240ms      8a6e4: MOVQ 0x30(SP), CX
      20ms       20ms      8a6e9: TESTB AL, 0(CX)
     250ms      250ms      8a6eb: MOVZX 0x1f(SP), DX
         .          .      8a6f0: MOVSX DL, DX
         .          .      8a6f4: CMPQ $0x6, DX
         .          .      8a6f8: JAE 0x8a721
         .          .      8a6fa: SHLQ $0xf, DX
      40ms       40ms      8a6fe: ADDQ CX, DX
         .          .      8a701: ANDL $0xfff, AX
         .          .      8a706: CMPQ $0x1000, AX
         .          .      8a70c: JAE 0x8a721
         .          .      8a70e: LEAQ 0(DX)(AX*8), AX
     100ms      100ms      8a712: MOVQ AX, 0x50(SP)
         .          .      8a717: MOVQ 0x20(SP), BP
         .          .      8a71c: ADDQ $0x28, SP
         .          .      8a720: RET
         .          .      8a721: CALL runtime.panicindex(SB)
         .          .      8a726: UD2
         .          .      8a728: CALL runtime.morestack_noctxt(SB)
         .          .      8a72d: JMP go.uber.org/zap/zapcore.(*counters).get(SB)
         .          .      8a732: INT $0x3
         .          .      8a733: INT $0x3
         .          .      8a734: INT $0x3
         .          .      8a735: INT $0x3
         .          .      8a736: INT $0x3
         .          .      8a737: INT $0x3
         .          .      8a738: INT $0x3
         .          .      8a739: INT $0x3
         .          .      8a73a: INT $0x3
         .          .      8a73b: INT $0x3
         .          .      8a73c: INT $0x3
         .          .      8a73d: INT $0x3
         .          .      8a73e: INT $0x3
```